

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; SimPrily 1.0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="SimPrily 1.0.1 documentation" href="index.html"/>
        <link rel="next" title="High Throughput Computing" href="htc.html"/>
        <link rel="prev" title="Install and Environment Set up" href="install.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> SimPrily
          

          
          </a>

          
            
            
              <div class="version">
                1.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Install and Environment Set up</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#required-input">Required Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optional-input">Optional Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#additional-information-on-input-arguments">Additional information on input arguments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id">ID</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-dir">output_dir</a></li>
<li class="toctree-l3"><a class="reference internal" href="#param-file-txt">param_file.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="#model-file-csv">model_file.csv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#setup-simulation-arguments"><strong>Setup simulation arguments</strong></a></li>
<li class="toctree-l4"><a class="reference internal" href="#demographic-simulation-arguments"><strong>Demographic simulation arguments</strong></a></li>
<li class="toctree-l4"><a class="reference internal" href="#snp-array-ascertainment-arguments"><strong>SNP array ascertainment arguments</strong></a></li>
<li class="toctree-l4"><a class="reference internal" href="#ordering-of-time-specific-events"><strong>Ordering of time-specific events</strong></a></li>
<li class="toctree-l4"><a class="reference internal" href="#germline"><strong>germline</strong></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pedmap"><strong>pedmap</strong></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="htc.html">High Throughput Computing</a></li>
<li class="toctree-l1"><a class="reference internal" href="real.html">Calculating summary statistics on real data</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_docker.html">Tutorial with Docker</a></li>
<li class="toctree-l1"><a class="reference internal" href="develop.html">For Developers</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SimPrily</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>For quick help:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simprily</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
<p>e.g. One simulation (with pseudo array and genetic map):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simprily</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">p</span> <span class="n">examples</span><span class="o">/</span><span class="n">eg1</span><span class="o">/</span><span class="n">param_file_eg1_asc</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">examples</span><span class="o">/</span><span class="n">eg1</span><span class="o">/</span><span class="n">model_file_eg1_asc</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">g</span> <span class="n">genetic_map_b37</span><span class="o">/</span><span class="n">genetic_map_GRCh37_chr1</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">macshs</span> <span class="o">-</span><span class="n">a</span> <span class="n">array_template</span><span class="o">/</span><span class="n">ill_650_test</span><span class="o">.</span><span class="n">bed</span> <span class="o">-</span><span class="n">i</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">v</span>
</pre></div>
</div>
<p>e.g. One simulation (genetic map, no pseudo array):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simprily</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">p</span> <span class="n">examples</span><span class="o">/</span><span class="n">eg1</span><span class="o">/</span><span class="n">param_file_eg1</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">m</span> <span class="n">examples</span><span class="o">/</span><span class="n">eg1</span><span class="o">/</span><span class="n">model_file_eg1</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">g</span> <span class="n">genetic_map_b37</span><span class="o">/</span><span class="n">genetic_map_GRCh37_chr1</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">macshs</span> <span class="o">-</span><span class="n">i</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_dir</span> <span class="o">-</span><span class="n">v</span>
</pre></div>
</div>
<hr class="docutils" />
<p><code class="docutils literal"><span class="pre">simprily.py</span></code> takes 4 required arguments and 2 optional arguments, and help, verbose, and profile options.</p>
<p>Run as</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simprily</span><span class="o">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">-</span><span class="n">p</span> <span class="n">PARAM</span> <span class="o">-</span><span class="n">m</span> <span class="n">MODEL</span> <span class="o">-</span><span class="n">i</span> <span class="n">ID</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUT</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span> <span class="n">MAP</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">a</span> <span class="n">ARRAY</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">profile</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="required-input">
<h2>Required Input<a class="headerlink" href="#required-input" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-p <var>PARAM</var></span></kbd></td>
<td>The location of the parameter file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-m <var>MODEL</var></span></kbd></td>
<td>The location of the model file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-i <var>ID</var></span></kbd></td>
<td>The unique identifier of the job</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-o <var>OUT</var></span></kbd></td>
<td>The location of the output directory</td></tr>
</tbody>
</table>
<p>or</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--param <var>PARAM</var></span></kbd></td>
<td>The location of the parameter file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--model <var>MODEL</var></span></kbd></td>
<td>The location of the model file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--id <var>ID</var></span></kbd></td>
<td>The unique identifier of the job</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--out <var>OUT</var></span></kbd></td>
<td>The location of the output directory</td></tr>
</tbody>
</table>
</div>
<div class="section" id="optional-input">
<h2>Optional Input<a class="headerlink" href="#optional-input" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-h</span></kbd></td>
<td>Shows a help message and exists</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span></kbd></td>
<td>Increase output verbosity. This includes 3 levels, <code class="docutils literal"><span class="pre">-v</span></code>, <code class="docutils literal"><span class="pre">-vv</span></code>, and <code class="docutils literal"><span class="pre">-vvv</span></code></td></tr>
<tr><td class="option-group">
<kbd><span class="option">--profile</span></kbd></td>
<td>Print a log file containing the time in seconds and memory use in Mb for main functions</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-g <var>MAP</var></span></kbd></td>
<td>The location of the genetic map file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-a <var>ARRAY</var></span></kbd></td>
<td>The location of the array template file, in <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/general-usage.html">bed format</a>. The third column is used as the physical positions of the SNP for the pseudo array.</td></tr>
</tbody>
</table>
<p>or</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--help</span></kbd></td>
<td>Shows a help message and exists</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span></kbd></td>
<td>Increase output verbosity. This includes 3 levels, <code class="docutils literal"><span class="pre">-v</span></code>, <code class="docutils literal"><span class="pre">-vv</span></code>, and <code class="docutils literal"><span class="pre">-vvv</span></code></td></tr>
<tr><td class="option-group">
<kbd><span class="option">--profile</span></kbd></td>
<td>Print a log file containing the time in seconds and memory use in Mb for main functions</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--map <var>MAP</var></span></kbd></td>
<td>The location of the genetic map file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--array <var>ARRAY</var></span></kbd></td>
<td>The location of the array template file, in <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/general-usage.html">bed format</a>. The third column is used as the physical positions of the SNP for the pseudo array.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="additional-information-on-input-arguments">
<h2>Additional information on input arguments<a class="headerlink" href="#additional-information-on-input-arguments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id">
<h3>ID<a class="headerlink" href="#id" title="Permalink to this headline">¶</a></h3>
<p>This is a unique identifier for the job. It is used in the names of the output files.
For example, the output file with parameter values and summary statistics is named <code class="docutils literal"><span class="pre">results_{IDid}.txt</span></code>.</p>
</div>
<div class="section" id="output-dir">
<h3>output_dir<a class="headerlink" href="#output-dir" title="Permalink to this headline">¶</a></h3>
<p>This is where all the output goes.
Within the output_dir the directory <code class="docutils literal"><span class="pre">results</span></code> will always be created. The <code class="docutils literal"><span class="pre">results</span></code> directory contains the results file <code class="docutils literal"><span class="pre">results_{jobid}.txt</span></code> with the parameter values and summary statistics.
Additionally, the directories <code class="docutils literal"><span class="pre">germline_out</span></code> and <code class="docutils literal"><span class="pre">sim_data</span></code> are also created, but will be empty if the <code class="docutils literal"><span class="pre">germline</span></code> or <code class="docutils literal"><span class="pre">pedmap</span></code> arguments in the model file are not included.</p>
<p><em>Be careful when running large numbers of jobs (&gt;2000). It is bad practice to run large numbers of jobs and direct all the output to the same directory, because listing the contents of the directory becomes very slow. Instead, we recommend creating directory “buckets”. See section Recommendations for other HTC workflows</em>.</p>
</div>
<div class="section" id="param-file-txt">
<h3>param_file.txt<a class="headerlink" href="#param-file-txt" title="Permalink to this headline">¶</a></h3>
<p>Examples of param_file.txt can be found in examples.
The param_file.txt must define the parameters of the demographic model and the minimum derived allele frequency to be used to create the pseudo array, if a pseudo array is to be created.</p>
<p>All time parameters must end in <code class="docutils literal"><span class="pre">_t</span></code>.</p>
<p>All parameter values should be given in pre-coalescent scaled units.
That is, Ne should be given in units of chromosomes, and time should be given in units of generations.
The code will scale to the appropriate coalescent units for the simulation.</p>
<p>The definition can be hard-coded parameter values, such as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">B</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">T1_t</span> <span class="o">=</span> <span class="mi">100</span>
</pre></div>
</div>
<p>The definition can be a prior, such as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1e3</span><span class="o">.</span><span class="mi">0</span><span class="p">:</span><span class="mf">1e4</span><span class="o">.</span><span class="mi">0</span><span class="p">)</span>
<span class="n">B</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1e3</span><span class="o">.</span><span class="mi">0</span><span class="p">:</span><span class="mf">1e4</span><span class="o">.</span><span class="mi">0</span><span class="p">)</span>
<span class="n">T1_t</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">:</span><span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
<p>Log base 10 can be used for the parameter definitions by using <code class="docutils literal"><span class="pre">1eX</span></code> or <code class="docutils literal"><span class="pre">1Ex</span></code>.
This is recommended when using a prior with a very large range (See ABCtoolbox manual).</p>
<p>If pseudo arrays are to be created, the derived allele frequency must be defined. For example,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1e3</span><span class="o">.</span><span class="mi">0</span><span class="p">:</span><span class="mf">1e4</span><span class="o">.</span><span class="mi">0</span><span class="p">)</span>
<span class="n">B</span> <span class="o">=</span> <span class="p">(</span><span class="mf">1e3</span><span class="o">.</span><span class="mi">0</span><span class="p">:</span><span class="mf">1e4</span><span class="o">.</span><span class="mi">0</span><span class="p">)</span>
<span class="n">T1_t</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">:</span><span class="mi">500</span><span class="p">)</span>
<span class="n">daf</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.01</span><span class="p">:</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<p><em>currently only a range of values is supported for daf. Therefore if you want to hard code a value, use the same value as the min and max of the prior.</em></p>
</div>
<div class="section" id="model-file-csv">
<h3>model_file.csv<a class="headerlink" href="#model-file-csv" title="Permalink to this headline">¶</a></h3>
<p>Examples of model_file.csv can be found in examples.</p>
<p>The demographic model, SNP ascertainment model, and additional options are defined in the model_file.csv.
The demographic model defines events in populations’ history, including population divergence, instantanious effective population size changes, exponential growth, gene flow and admixture. We use a coalescent simulation, so models must be defined backwards in time, starting from the present, with each event going back in the past. The SNP ascertainment model defines how to create a pseudo SNP array using a template SNP array, a set of discovery populations and a minor allele frequency cutoff. The SNP ascertainment model should be used when comparing to real SNP array data.</p>
<p>All instances of any argument must start with a dash followed by the corresponding argument parameters,
and value(s).
Each new argument must be a new line.
All variables and values must be separated by commas (white space will be ignored, so it is okay to include spaces).
The model arguments can appear in any order.</p>
<p>All parameters must be called with a name corresponding to its definition in the param file.
This is how parameter values are assigned to the simulation model.
For example,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">macs</span><span class="p">,</span><span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">macs</span><span class="p">,</span>
<span class="o">-</span><span class="n">length</span><span class="p">,</span><span class="mi">5000000</span><span class="p">,</span>
<span class="o">-</span><span class="n">s</span><span class="p">,</span><span class="mi">1231414</span><span class="p">,</span>
<span class="o">-</span><span class="n">t</span><span class="p">,</span><span class="mf">2.5e-8</span><span class="p">,</span>
<span class="o">-</span><span class="n">r</span><span class="p">,</span><span class="mf">1e-8</span><span class="p">,</span>
<span class="o">-</span><span class="n">h</span><span class="p">,</span><span class="mf">1e5</span><span class="p">,</span>
<span class="c1"># define a sample size of 50 haploid individuals for populations 1 and 2</span>
<span class="o">-</span><span class="n">I</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">50</span>
<span class="c1"># define the effective population size at present for population 1</span>
<span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">A</span>
<span class="c1"># define the effective population size at present for population 2</span>
<span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">B</span>
<span class="c1"># define a divergence event (join backwards in time) between populations 1 and 2</span>
<span class="o">-</span><span class="n">ej</span><span class="p">,</span> <span class="n">T1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span>
</pre></div>
</div>
<div class="section" id="setup-simulation-arguments">
<h4><strong>Setup simulation arguments</strong><a class="headerlink" href="#setup-simulation-arguments" title="Permalink to this headline">¶</a></h4>
<p>One of the following two flags must be included:</p>
<dl class="docutils">
<dt><code class="docutils literal"><span class="pre">-macs</span></code></dt>
<dd>use the original simulator <a class="reference external" href="https://github.com/gchen98/macs">MaCS</a>. This option will stream the MaCS simulation output directly to be read into a python bitarray.</dd>
<dt><code class="docutils literal"><span class="pre">-macs_file</span></code></dt>
<dd>read in static output from MaCS. This should only be used for rigorous testing.</dd>
</dl>
<p>Following the <code class="docutils literal"><span class="pre">-macs</span></code> and <code class="docutils literal"><span class="pre">-macs_file</span></code> flags there should be a path to either the executable or static file in relation to the working directory. For example:</p>
<blockquote>
<div><p>If you are using a virtual environment the path to macs should be</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">macs</span><span class="p">,</span> <span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">macs</span>
</pre></div>
</div>
<p>If you are using Docker or Singularity the path to macs should be</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">macs</span><span class="p">,</span> <span class="o">/</span><span class="n">app</span><span class="o">/</span><span class="n">macs</span>
</pre></div>
</div>
<p>or if you want to use a static file,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">macs_file</span><span class="p">,</span> <span class="n">tests</span><span class="o">/</span><span class="n">test_data</span><span class="o">/</span><span class="n">sites1000000</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
</div></blockquote>
<dl class="docutils">
<dt><code class="docutils literal"><span class="pre">-length</span></code></dt>
<dd>The number base pairs you want to simulate. Must be included.</dd>
<dt><code class="docutils literal"><span class="pre">-s</span></code></dt>
<dd>random seed.
Must be an integer.
If no input is given, no seed will be used, and everything will be random.
If a seed is provided, reproducible parameters will be picked from the priors.
Using a seed will also cause reproducible simulations with macs.</dd>
</dl>
</div>
<div class="section" id="demographic-simulation-arguments">
<h4><strong>Demographic simulation arguments</strong><a class="headerlink" href="#demographic-simulation-arguments" title="Permalink to this headline">¶</a></h4>
<p>All argument flags are based on macs arguments (see macs and ms manual for more detail).</p>
<p><code class="docutils literal"><span class="pre">-t</span></code>: mutation rate per site per 4N generations</p>
<p><code class="docutils literal"><span class="pre">-d</span></code>: enable debugging messages. No entry will default to allowing debugging messages. This will not work when using macsswig</p>
<p><code class="docutils literal"><span class="pre">-h</span></code>: history. Refers to the number of previous base pairs to retain</p>
<p><code class="docutils literal"><span class="pre">-r</span> <span class="pre">[r]</span></code>:  recombination rate per site per 4N generations</p>
<p><code class="docutils literal"><span class="pre">-c</span> <span class="pre">[f</span> <span class="pre">lambda]</span></code>: f = ratio of gene conversion rate to crossover rate. track len(lambda) is mean length of tract in base pairs.
<em>This has not been tested.</em></p>
<p><code class="docutils literal"><span class="pre">-T</span></code>: Print each local tree in Newick format to standard out. <em>This has not been tested.</em></p>
<p><code class="docutils literal"><span class="pre">-G</span> <span class="pre">[alpha]</span></code>: Assign growth rate alpha across populations where alpha=-log(Np/Nr).</p>
<p><code class="docutils literal"><span class="pre">-I</span> <span class="pre">[n</span> <span class="pre">n_n]</span></code>: Assign all elements of the migration matrix for n populations.
Values in matrix set to mig_rate/(n-1).
The length of n_n should be equal to n</p>
<p><code class="docutils literal"><span class="pre">-m</span> <span class="pre">[i,j</span> <span class="pre">m]</span></code>: i, j is associated with a location in the migration matrix
m is assigned to the value at (i, j)</p>
<p><code class="docutils literal"><span class="pre">-ma</span> <span class="pre">[m_nn]</span></code>: Assign values to all elements of migration matrix for n populations</p>
<p><code class="docutils literal"><span class="pre">-n</span> <span class="pre">[i</span> <span class="pre">size]</span></code>: Population i set to size</p>
<p><code class="docutils literal"><span class="pre">-g</span> <span class="pre">[i</span> <span class="pre">alpha]</span></code>: assigns alpha value as explained in -G to population i</p>
<p><code class="docutils literal"><span class="pre">-eG</span> <span class="pre">[t</span> <span class="pre">alpha]</span></code>: t is a time value.
alpha behaves the same as in -G</p>
<p><code class="docutils literal"><span class="pre">-eg</span> <span class="pre">[t</span> <span class="pre">i</span> <span class="pre">alpha]</span></code>:
t is a time value.
alpha behaves the same as in -G.
i is a population that alpha is assigned to at time t.</p>
<p><code class="docutils literal"><span class="pre">-eM</span> <span class="pre">[t</span> <span class="pre">m]</span></code>:
t is a time value.
Assign migration rate m to all elements in migration matrix at
time t</p>
<p><code class="docutils literal"><span class="pre">-em</span> <span class="pre">[t</span> <span class="pre">i,j</span> <span class="pre">m_ij]</span></code>:
t is a time value.
i and j make up point in a population matrix.
assigns migration rate m_ij to the population at i, j at time t</p>
<p><code class="docutils literal"><span class="pre">-ema</span> <span class="pre">[t</span> <span class="pre">n</span> <span class="pre">m_nn]</span></code>:
t is a time value.
Assign migration rates  within the migration matrix for n
populations at time t.</p>
<p><code class="docutils literal"><span class="pre">-eN</span> <span class="pre">[t</span> <span class="pre">size]</span></code>:
t is a time value.
Assigns size to all populations at time t</p>
<p><code class="docutils literal"><span class="pre">-en</span> <span class="pre">[t</span> <span class="pre">i</span> <span class="pre">size_i]</span></code>:
t is a time value.
assigns size_i to population i at time t</p>
<p><code class="docutils literal"><span class="pre">-es</span> <span class="pre">[t</span> <span class="pre">i</span> <span class="pre">p]</span></code>:
t is a time value.
splits population i by p at time t</p>
<p><code class="docutils literal"><span class="pre">-ej</span> <span class="pre">[t</span> <span class="pre">i</span> <span class="pre">j]</span></code>
t is a time value.
joins population i with population j at time t</p>
</div>
<div class="section" id="snp-array-ascertainment-arguments">
<h4><strong>SNP array ascertainment arguments</strong><a class="headerlink" href="#snp-array-ascertainment-arguments" title="Permalink to this headline">¶</a></h4>
<p>If the user would like to create a pseudo array from the simulation, the array template must be included in the command line argument with the flag <code class="docutils literal"><span class="pre">-a</span></code>, and four additional arguments must be included in the model_file:</p>
<p><code class="docutils literal"><span class="pre">-discovery</span></code>, followed by the populations (defined by their numbers from <code class="docutils literal"><span class="pre">-n</span></code>) that should be used to discover the SNP (e.g. the HapMap populations).
These are the populations that will be used to create the pseudo array.
When calculating summary statistics, summary statistics based on whole genome simulation and pseudo array will be calculated for these populations.</p>
<p><code class="docutils literal"><span class="pre">-sample</span></code>, followed by the populations (defined by their numbers from <code class="docutils literal"><span class="pre">-n</span></code>) that are the samples of interest for demographic interest.</p>
<p><code class="docutils literal"><span class="pre">-daf</span></code>, followed by the parameter name for daf.</p>
<p><code class="docutils literal"><span class="pre">-random_discovery</span></code>, followed by <code class="docutils literal"><span class="pre">True</span></code> or <code class="docutils literal"><span class="pre">False</span></code>.
True will add a random number of individuals to the discovery populations to use as the “panel” to create the pseudo array.
When this option is False, the total number of simulated discovery populations is equal to the number “genotyped” and in the “panel”.</p>
<p>For example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">macs</span><span class="p">,</span><span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">macs</span><span class="p">,</span>
<span class="o">-</span><span class="n">length</span><span class="p">,</span><span class="mi">5000000</span><span class="p">,</span>
<span class="o">-</span><span class="n">s</span><span class="p">,</span><span class="mi">1231414</span><span class="p">,</span>
<span class="o">-</span><span class="n">t</span><span class="p">,</span><span class="mf">2.5e-8</span><span class="p">,</span>
<span class="o">-</span><span class="n">r</span><span class="p">,</span><span class="mf">1e-8</span><span class="p">,</span>
<span class="o">-</span><span class="n">h</span><span class="p">,</span><span class="mf">1e5</span><span class="p">,</span>
<span class="o">-</span><span class="n">I</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">50</span>
<span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">A</span>
<span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">B</span>
<span class="o">-</span><span class="n">ej</span><span class="p">,</span> <span class="n">T1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span>
<span class="o">-</span><span class="n">discovery</span><span class="p">,</span> <span class="mi">1</span>
<span class="o">-</span><span class="n">sample</span><span class="p">,</span> <span class="mi">2</span>
<span class="o">-</span><span class="n">daf</span><span class="p">,</span> <span class="n">daf</span>
<span class="o">-</span><span class="n">random_discovery</span><span class="p">,</span> <span class="kc">True</span>
</pre></div>
</div>
<p>An example of an array template is:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">15929526</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">15991515</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">16288162</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">16926611</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">16990146</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">17498992</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">17540297</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">17728199</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">17760714</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">18180154</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">18217275</span>
<span class="n">chr22</span>       <span class="mi">0</span>       <span class="mi">18220413</span>
</pre></div>
</div>
</div>
<div class="section" id="ordering-of-time-specific-events">
<h4><strong>Ordering of time-specific events</strong><a class="headerlink" href="#ordering-of-time-specific-events" title="Permalink to this headline">¶</a></h4>
<p>When using priors, if some demographic events must happen in a certain order, the order can be specified by adding the order number to the argument.
For example say there are two demographic events, a population split and instantaneous growth, but the instantaneous growth must happen before the population split, we can indicate that in the model file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">en_1</span><span class="p">,</span> <span class="n">Tgrowth</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">A2</span>
<span class="o">-</span><span class="n">ej_2</span><span class="p">,</span> <span class="n">Tsplit</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span>
</pre></div>
</div>
<p>Additionally, the same format can be used to indicate that multiple events should happen at the same time.
If there are multiple events that should happen at the same time, the word <code class="docutils literal"><span class="pre">inst</span></code> should be used instead of a time parameter after the first definition of the time.
<em>(this will actually cause the times to be just different enough that macs is happy.)</em>
For example, say we wanted growth to occur at the same time as the population split:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">en_1</span><span class="p">,</span> <span class="n">Tgrowth</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">A2</span>
<span class="o">-</span><span class="n">ej_1</span><span class="p">,</span> <span class="n">inst</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span>
</pre></div>
</div>
<p>In this case, the population split will technically be simulated slightly after the growth.</p>
</div>
<div class="section" id="germline">
<h4><strong>germline</strong><a class="headerlink" href="#germline" title="Permalink to this headline">¶</a></h4>
<p><em>currently has a bug</em></p>
<p>The option <code class="docutils literal"><span class="pre">-germline</span></code> can be included in the model_file to use <a class="reference external" href="https://github.com/sgusev/GERMLINE">GERMLINE</a> to find shared IBD segments between all simulated individuals from pseudo array.
Does not use the genetic map to run GERMLINE.
Runs GERMLINE as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bash</span> <span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">phasing_pipeline</span><span class="o">/</span><span class="n">gline</span><span class="o">.</span><span class="n">sh</span> <span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">germline</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="mi">5</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">germline</span>  <span class="n">ped_name</span> <span class="n">map_name</span> <span class="n">out_name</span> <span class="s2">&quot;-bits 10 -min_m min_m&quot;</span>
</pre></div>
</div>
<p>If GERMLINE does not run, try rebuilding it on the machine you are trying to run on:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">./</span><span class="nb">bin</span><span class="o">/</span><span class="n">germline</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="mi">5</span><span class="o">-</span><span class="mi">1</span>
<span class="n">make</span> <span class="n">clean</span>
<span class="n">make</span>
</pre></div>
</div>
</div>
<div class="section" id="pedmap">
<h4><strong>pedmap</strong><a class="headerlink" href="#pedmap" title="Permalink to this headline">¶</a></h4>
<p>The option <code class="docutils literal"><span class="pre">-pedmap</span></code> can be included in the model_file to print a ped and map file of the pseudo array data.</p>
</div>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="htc.html" class="btn btn-neutral float-right" title="High Throughput Computing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="install.html" class="btn btn-neutral" title="Install and Environment Set up" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Ariella Gladstein.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>